### PR TITLE
Add recipe for OpenMM-XTB

### PR DESCRIPTION
**Pull request by peastman**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

@conda-forge/help-python-c, ready for review!

@RaulPPelaez I included you as a maintainer.  I hope that's ok?

### CODE REVIEW

**Comment by RaulPPelaez** [comment]
thats ok

**Comment by RaulPPelaez** [comment]
I can help with the recipe but I do not have push access into this PR.
The build script should be simple, like this one:
```shell
#!/bin/bash

set -euxo pipefail

rm -rf build || true


CMAKE_FLAGS="${CMAKE_ARGS} -DCMAKE_INSTALL_PREFIX=${PREFIX} -DCMAKE_BUILD_TYPE=Release"
CMAKE_FLAGS+=" -DPython3_EXECUTABLE=${PYTHON}"
CMAKE_FLAGS+=" -DOPENMM_DIR=${PREFIX}"


mkdir build && cd build
cmake ${CMAKE_FLAGS} ${SRC_DIR}
make -j$CPU_COUNT
make -j$CPU_COUNT install
```

**Comment by peastman** [comment]
If you want to make changes yourself, you can create a PR to my branch and I'll merge them.

**Comment by RaulPPelaez** [comment]
The contents of the recipe folder translate directly t the final feedstock AFAIK. So that should go to recipe/build.sh.
Let me PR to your fork then.

**Comment by peastman** [comment]
I think the script also needs to `make PythonInstall`?  In addition, I see the [OpenMM-PLUMED build script](https://github.com/conda-forge/openmm-plumed-feedstock/blob/main/recipe/build.sh) packages up the tests.  Should this one do the same?

**Comment by xhochy** [comment]
Please don't request a review before this is ready for review.

**Comment by RaulPPelaez** [comment]
xtb is not built for windows. So that should be skipped in this package too
https://github.com/conda-forge/xtb-feedstock/blob/98a2f6893733e62bb411be616fd46503f3d424a9/recipe/meta.yaml#L14

**Comment by RaulPPelaez** [comment]
In OSX the lib is successfully built and installed, but then the test does not find the dynamic library:
`['  ERROR (openmm-xtb,share/openmm-xtb/tests/TestXtbForce): /Users/runner/Miniforge3/conda-bld/openmm-xtb_1707754971797/work/build/libOpenMMXTB.dylib not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?']`

I do not know what to do about that.

**Comment by peastman** [comment]
Earlier in the log it says,

> 2024-02-12T16:25:23.2315870Z -- Installing: $PREFIX/lib/libOpenMMXTB.dylib

So it clearly is being built.  Perhaps it's looking in the wrong location?  A few lines before the error, there's the line

> 2024-02-12T16:25:45.4007130Z    INFO (openmm-xtb,lib/python3.10/site-packages/_openmmxtb.cpython-310-darwin.so): lib/libOpenMMXTB.dylib found in this package

I think that means the Python package successfully linked to it.  But when it tries to run `TestXtbForce`, it looks in a different location.

**Comment by RaulPPelaez** [comment]
Maybe some rpath play is required here?

```yaml
  rpaths:
    - lib/

```

**Comment by peastman** [comment]
The documentation says that's a Linux feature, and it's ignored on other platforms.

**Comment by peastman** [comment]
Possibly that error message is misleading and the real problem is something different.  Earlier in the log we get the warning

> WARNING (openmm-xtb): interpreter (Python) package conda-forge/osx-64::python==3.10.13=h00d2728_1_cpython in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)

And then just before the error about TextXtbForce there's an exception

> conda_build.exceptions.OverLinkingError: overlinking check failed

[The documentation](https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#ignoring-files-in-overlinking-overdepending-checks) briefly mentions options related to overlinking and overdepending, but I can't find any definition of what those terms mean.

It also seems like it could be an issue related to relocatable binaries: it's looking for the library in the environment where it was built instead of where it's being tested.  But the documentation says binaries are made relocatable by default.

**Comment by RaulPPelaez** [comment]
I see a million of those warnings in all feedstocks. I always assumed it was an overprotecting test. As in, it is not correctly understanding what "used" is.
Perhaps one has to use ${PYTHON} instead of python when testing?

**Comment by RaulPPelaez** [comment]
What about silencing the error with:
```yaml
build:
   missing_dso_whitelist:
      - libOpenMMXTB.dylib
```

**Comment by peastman** [comment]
Success!

**Comment by peastman** [comment]
This is now working.  It's ready to be reviewed and merged.

**Comment by RaulPPelaez** [comment]
cc @conda-forge/help-python-c could you please take a look? Thanks.

File: recipes/openmm-xtb/meta.yaml
Comment (xhochy) [inline recipes/openmm-xtb/meta.yaml:26]: This should never be listed as a dependency.
```suggestion
```

File: recipes/openmm-xtb/meta.yaml
Comment (xhochy) [inline recipes/openmm-xtb/meta.yaml:35]: Aren't the libraries using a run_exports? Normally you shouldn't need to do this for native libraries.

File: recipes/openmm-xtb/build.sh
Comment (xhochy) [inline recipes/openmm-xtb/build.sh:23]: Typically, we don't install tests. Why are you doing it here?

File: recipes/openmm-xtb/build.sh
Comment (xhochy) [inline recipes/openmm-xtb/build.sh:5]: This line is not actually used.

**Review by xhochy — CHANGES_REQUESTED** [review]

File: recipes/openmm-xtb/meta.yaml
Comment (peastman) [reply recipes/openmm-xtb/meta.yaml:35]: Apparently not.  There's no `run_exports` in https://github.com/conda-forge/openmm-feedstock/blob/main/recipe/meta.yaml.

File: recipes/openmm-xtb/build.sh
Comment (peastman) [reply recipes/openmm-xtb/build.sh:23]: @jaimergp this is copied from the build script you wrote for openmm-plumed.  Do you remember what the reason was?

Comment (peastman) [reply recipes/openmm-xtb/build.sh:20]: @xhochy is there a way to run the tests without copying them to the package?  I looked up a few other packages to see how they handled tests, but they did exactly the same thing as here.

According to the [conda-build documentation](https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#test-section) you can add a `source_files` section to the recipe to make it copy files to the directory where the tests will be run.  I tried adding

```yaml
test:
  source_files:
    - python/tests/TestXtbForce.py
```

But when I then try to run it with `python TestXtbForce.py`, it says the file isn't there.

Comment (jaimergp) [reply recipes/openmm-xtb/build.sh:20]: This was handled this way in Plumed because they were only four files and it was easier that way. IIRC, the test files also had the prefix hardcoded and tests would fail otherwise. It was also the only way to test reliably whether the package was correctly built.

`files` and `source_files` will end up "packaging" the files too in the sense that they will be published as part of the `info/` directory, although maybe not part of the PREFIX, if that's what we are concerned with.

> But when I then try to run it with python TestXtbForce.py, it says the file isn't there.

Try `python python/tests/TestXtbForce.py`. They are copied with the relative paths untouched.

Comment (peastman) [reply recipes/openmm-xtb/build.sh:20]: The situation is the same here.  There are only two files (one C++ test and one Python test), both very small.  If `source_files` also adds it to the package, I don't see any advantage to doing it that way?

File: recipes/openmm-xtb/meta.yaml
Comment (peastman) [reply recipes/openmm-xtb/meta.yaml]: Fixed.

File: recipes/openmm-xtb/build.sh
Comment (peastman) [reply recipes/openmm-xtb/build.sh]: Fixed.

File: recipes/openmm-xtb/meta.yaml
Comment (peastman) [reply recipes/openmm-xtb/meta.yaml:26]: Actually, removing it broke the Mac build.

```
2024-03-06T01:30:42.7708550Z Traceback (most recent call last):
2024-03-06T01:30:42.7710240Z   File "/Users/runner/work/1/s/recipes/openmm-xtb/patch_osx_tests.py", line 4, in <module>
2024-03-06T01:30:42.7710700Z     from conda_build.os_utils import macho
2024-03-06T01:30:42.7711150Z ModuleNotFoundError: No module named 'conda_build'
```

**Comment by peastman** [comment]
@conda-forge/help-python-c this is ready for another review.

**Comment by peastman** [comment]
Can someone review this?

**Comment by peastman** [comment]
If no one else has any comments, can someone merge this?

File: recipes/openmm-xtb/meta.yaml
Comment (jaimergp) [reply recipes/openmm-xtb/meta.yaml:26]: That error comes from the test helper script, where we had to patch some test binaries so they don't contain the hardcoded prefix. I'm not sure this is needed in this project (again, that was a plumed thing). I'd start simple with more standard approaches first, and then add hacks as needed using the existing openmm-* recipes as hints.

File: recipes/openmm-xtb/build.sh
Comment (jaimergp) [inline recipes/openmm-xtb/build.sh:26]: Can we check if we still need this?

Comment (peastman) [reply recipes/openmm-xtb/build.sh]: It's still needed.  WIthout it we get the `conda_build.exceptions.OverLinkingError: overlinking check failed` error.

File: recipes/openmm-xtb/meta.yaml
Comment (peastman) [reply recipes/openmm-xtb/meta.yaml:26]: As noted in my other comment, it really is needed.  I did exactly what you described: started with the simple approach, then copied over the fixes for problems from other recipes.

File: recipes/openmm-xtb/build.sh
Comment (jaimergp) [reply recipes/openmm-xtb/build.sh:26]: For future reference, this is the full error:

```
  ERROR (openmm-xtb,share/openmm-xtb/tests/TestXtbForce): /Users/runner/Miniforge3/conda-bld/openmm-xtb_1712618910781/work/build/libOpenMMXTB.dylib not found in packages, sysroot(s) nor the missing_dso_whitelist.
```

Where we can see that the `build_dir` path is hardcoded so it doesn't end up being replaced by the PREFIX machinery in conda-build. To me this sounds like something that could be fixed in the CMake config, preferrably.

That said I don't know exactly where that'd go so I'd support this workaround for now.

@xhochy, what do you think?

Comment (xhochy) [reply recipes/openmm-xtb/build.sh:26]: Sounds like a good workaround. We have tried to fix  this in (upstream) Arrow, too, but we never found a reliable solution :(

Comment (peastman) [reply recipes/openmm-xtb/build.sh:26]: Thanks!  Is it ok to merge this?

Comment (xhochy) [reply recipes/openmm-xtb/build.sh:26]: From my PoV yes. I'll let @jaimergp have a final look.

**Review by jaimergp — APPROVED** [review]

**Pull request merged**
